quakelib/src/QuakeLibIO.cpp

```diff
@@ -2843,6 +2847,141 @@ int quakelib::ModelEventSet::read_file_ascii(const std::string &event_file_name,
+    size_t                      *field_offsets;
+    size_t                      *field_sizes;
+    herr_t                      res;
+
+    _events.clear();
+    descs.clear();
+    ModelEvent::get_field_descs(descs);
+    num_fields = descs.size();
+    field_offsets = new size_t[num_fields];
+    field_sizes = new size_t[num_fields];
+
+    for (i=0; i<num_fields; ++i) {
+        field_offsets[i] = descs[i].offset;
+        field_sizes[i] = descs[i].size;
+    }
+
+    res = H5TBget_table_info(data_file, ModelEvent::hdf5_table_name().c_str(), &num_fields, &num_events);
+
+    if (res < 0) exit(-1);
+
+    // TODO: check that num_fields matches the descs
+
+    event_data = new EventData[num_events];
+    res = H5TBread_records(data_file, ModelEvent::hdf5_table_name().c_str(), 0, num_events, sizeof(EventData), field_offsets, field_sizes, event_data);
+
+    if (res < 0) exit(-1);
+
+    // Read section data into the World
+    for (i=0; i<num_events; ++i) {
+        ModelEvent  new_event;
+        new_event.read_data(event_data[i]);
+        _events.push_back(new_event);
+    }
+
+    // Free memory for HDF5 related data
+    delete event_data;
+    delete field_offsets;
+    delete field_sizes;
+#else
+    // TODO: Error out
+#endif
+}
+
+void quakelib::ModelEventSet::read_sweeps_hdf5(const hid_t &data_file) {
+#ifdef HDF5_FOUND
+    std::vector<FieldDesc>                          descs;
+    ModelEventSet::iterator                   fit;
+    hsize_t                     num_fields, num_sweeps;
+    unsigned int                i;
+    unsigned int                start_sweep;
+    unsigned int                end_sweep;
+    SweepData                   *event_sweeps;
+    size_t                      *field_offsets;
+    size_t                      *field_sizes;
+    herr_t                      res;
+    
+    descs.clear();
+    ModelSweeps::get_field_descs(descs);
+    num_fields = descs.size();
+    field_offsets = new size_t[num_fields];
+    field_sizes = new size_t[num_fields];
+
+    for (i=0; i<num_fields; ++i) {
+        field_offsets[i] = descs[i].offset;
+        field_sizes[i] = descs[i].size;
+    }
+
+    res = H5TBget_table_info(data_file, ModelSweeps::hdf5_table_name().c_str(), &num_fields, &num_sweeps);
+
+    if (res < 0) exit(-1);
+
+    event_sweeps = new SweepData[num_sweeps];
+    res = H5TBread_records(data_file, ModelSweeps::hdf5_table_name().c_str(), 0, num_sweeps, sizeof(SweepData), field_offsets, field_sizes, event_sweeps);
+    
+    if (res < 0) exit(-1);
+    
+    // Read sweeps data into the ModelEventSet
+    for (fit=_events.begin(); fit!=_events.end(); ++fit) {
+        fit->getStartEndSweep(start_sweep, end_sweep);
+        ModelSweeps new_sweeps;
+        
+        for (i=start_sweep; i<end_sweep; i++) {
+            new_sweeps.read_data(event_sweeps[i]);
+        }
+        fit->setSweeps(new_sweeps);
+        
+    }
+#else
+    // TODO: Error out
+#endif
+    delete event_sweeps;
```

Comment:
This delete should be just above the #else above, otherwise the non-HDF5 case will delete a NULL pointer and it will crash.  Thanks!